client-apis/decrypt.md

```diff
@@ -69,65 +91,226 @@ This default CMM MUST obtain the decryption materials required for decryption.
-  Refer to [Message Body AAD](../data-format/message-body-aad.md) specification for more information.
-- Sequence Number: This value is the [sequence number](../data-format/message-body.md#sequence-number) of the frame being decrypted, if the message contains framed data.
-  If the message contains non framed data, then this value is 1.
-- Content Length: TODO
+The caller MAY forgoe indicating an end to the encrypted message.
+This operation MAY still succeed as long as it can successfully process the first
+sequential bytes inputted to it as an encrypted message.
 
-If the algorithm suite has a signature algorithm, decrypt MUST verify the message footer using the specified signature algorithm,
-by using the verification key obtained from the decryption materials.
+If any of these steps fails, this oepration MUST falt and indicate a failure to the caller.
 
-## Security Considerations
+### Parse the header
+
+Given encrypted message bytes, this operation MUST process those bytes sequentially,
+deserializing those bytes according to the [message format](../data-format/message.md).
+
+This operation MUST attempt to deserialize all available encrypted message bytes until it has
+successfully deserialized a valid [message header](../data-format/message-header.md).
+
+This operation MUST wait if it doesn't have enough encrypted message bytes available to
+deserialize the next field of the message header until enough input bytes become available or
+the caller indicates an end to the encrypted message.
 
-[TODO]
+Once a valid message header is deserialized,
+this operation MUST validate the [message header body](../data-format/message-header.md#header-body)
+by using the [authenticated encryption algorithm](../framework/algorithm-suites.md#encryption-algorithm)
+to decrypt with the following inputs:
 
-## Apendix
+- the AAD is the serialized [message header body](../data-format/message-header.md#header-body).
+- the IV has a value of 0
+- the cipherkey is the derived data key
+- the ciphertext is an empty byte array
+
+If this decryption fails, this operation MUST immediately halt and fail.
+
+If the algorithm suite contains a signature algorithm and
+this operation is streaming the plaintext output to the caller,
+this operation MUST input the serialized frame to the signature algorithm as soon as it is deserialized,
+such that the serialized frame isn't required to remain in memory to complete
+the [signature calculation](#signature-calculation).
+
+If this decryption succeeds and the algorithm suite does not contain a signature algorithm,
+the parsed [encryption context](#encryption-context) MAY be released to the caller.
+
+If the algorithm suite does contain a signature algorithm
+the parsed [encryption context](#encryption-context) MUST NOT be released to the caller until the
+[signature is verified](#verify-the-signature-optional).
```

Comment:
I thought we determined it could be released, just like we release the plaintext prior to signature verification.